File: api/py/mpa/process.py
"""
Monthly Performance Analysis - Process Endpoint

Vercel Python Function that runs the full MPA analysis pipeline.

POST /api/mpa/process
Body: { "batchId": "uuid" }

The batch must already exist in mpa_analysis_batches with file paths populated.
This function:
1. Downloads files from Supabase Storage
2. Runs the analysis pipeline
3. Saves results to database
4. Updates batch status
"""

import json
import os
import sys
import traceback
from http.server import BaseHTTPRequestHandler
from io import BytesIO

# Add lib directory to path for imports
current_dir = os.path.dirname(os.path.abspath(__file__))
lib_dir = os.path.join(current_dir, 'lib')
if lib_dir not in sys.path:
    sys.path.insert(0, lib_dir)

# Load environment from .env if running locally
try:
    from dotenv import load_dotenv
    load_dotenv()
except ImportError:
    pass

from loaders import (
    ProFormaLoader,
    CompensationLoader,
    HarvestHoursLoader,
    HarvestExpensesLoader,
    PnLLoader,
)
from classification import ProjectClassifier, classify_all_activity
from computations import (
    calculate_labor_costs,
    calculate_expense_costs,
    merge_direct_costs,
    calculate_cost_center_costs,
    calculate_non_revenue_client_costs,
)
from allocations import OverheadAllocator, calculate_margins, get_tagged_revenue
from validators import run_all_validations
from db import SupabaseClient


def run_analysis(db: SupabaseClient, batch: dict) -> dict:
    """
    Run the full MPA analysis pipeline.

    Args:
        db: SupabaseClient instance
        batch: Batch record from database

    Returns:
        Dictionary with analysis results and logs
    """
    batch_id = batch['id']
    month = batch['month_name']
    logs = []

    # Phase 1: Load files
    logs.append(f"Loading files for {month}...")

    proforma_content = db.download_file(batch['proforma_file_path'])
    compensation_content = db.download_file(batch['compensation_file_path'])
    hours_content = db.download_file(batch['hours_file_path'])
    expenses_content = db.download_file(batch['expenses_file_path'])
    pnl_content = db.download_file(batch['pnl_file_path'])

    # Load Pro Forma
    proforma_loader = ProFormaLoader(proforma_content, month)
    proforma_df = proforma_loader.load()
    logs.extend(proforma_loader.logs)

    # Load Compensation
    comp_loader = CompensationLoader(compensation_content)
    comp_df = comp_loader.load()
    logs.extend(comp_loader.logs)

    # Load Hours
    hours_loader = HarvestHoursLoader(hours_content, month)
    hours_df = hours_loader.load()
    logs.extend(hours_loader.logs)

    # Load Expenses
    expenses_loader = HarvestExpensesLoader(expenses_content)
    expenses_df = expenses_loader.load()
    logs.extend(expenses_loader.logs)

    # Load P&L
    pnl_loader = PnLLoader(pnl_content)
    pnl_df = pnl_loader.load()
    logs.extend(pnl_loader.logs)

    logs.append("Files loaded successfully")

    # Phase 2: Classify projects
    logs.append("Classifying projects...")
    classifier = ProjectClassifier()
    classified = classify_all_activity(proforma_df, hours_df, expenses_df, classifier)
    revenue_centers = classified['revenue_centers']
    cost_centers = classified['cost_centers']
    non_revenue_clients = classified['non_revenue_clients']

    logs.append(
        f"Classified: {len(revenue_centers)} revenue centers, "
        f"{len(cost_centers)} cost centers, "
        f"{len(non_revenue_clients)} non-revenue clients"
    )

    # Phase 3: Compute direct costs
    logs.append("Computing direct costs...")
    labor_summary, hours_detail, labor_logs = calculate_labor_costs(hours_df, comp_df)
    logs.extend(labor_logs)

    expense_summary, expense_detail = calculate_expense_costs(expenses_df)

    revenue_centers = merge_direct_costs(revenue_centers, labor_summary, expense_summary)
    cost_centers = calculate_cost_center_costs(cost_centers, hours_detail, expense_detail)
    non_revenue_clients = calculate_non_revenue_client_costs(non_revenue_clients, hours_detail, expense_detail)

    logs.append("Direct costs computed")

    # Phase 4: Allocate overhead
    logs.append("Allocating overhead pools...")
    allocator = OverheadAllocator()
    pools = allocator.calculate_pools(pnl_df, cost_centers, include_cc_in_sga=True)

    revenue_centers = allocator.allocate_sga(revenue_centers, pools['sga_pool'])
    revenue_centers = allocator.allocate_data(revenue_centers, pools['data_pool'])
    revenue_centers = allocator.allocate_workplace(revenue_centers, pools['workplace_pool'])
    revenue_centers = calculate_margins(revenue_centers)

    tagged_revenue = get_tagged_revenue(revenue_centers)

    logs.append(
        f"Pools allocated: SG&A ${pools['sga_pool']:,.2f}, "
        f"Data ${pools['data_pool']:,.2f}, "
        f"Workplace ${pools['workplace_pool']:,.2f}"
    )

    # Phase 5: Validate
    logs.append("Running validation checks...")
    validation_data = {
        'revenue_centers': revenue_centers,
        'cost_centers': cost_centers,
        'non_revenue_clients': non_revenue_clients,
        'proforma': proforma_df,
        'pools': pools,
        'hours': hours_df,
        'expenses': expenses_df,
        'compensation': comp_df,
        'pnl': pnl_df,
    }
    validation_results = run_all_validations(validation_data)
    logs.append(f"Validation: {validation_results.summary()}")

    # Calculate summary metrics
    total_revenue = float(revenue_centers['revenue'].sum())
    total_labor_cost = float(revenue_centers['labor_cost'].sum())
    total_expense_cost = float(revenue_centers['expense_cost'].sum())
    total_margin_dollars = float(revenue_centers['margin_dollars'].sum())
    overall_margin_percent = (total_margin_dollars / total_revenue * 100) if total_revenue > 0 else 0

    summary = {
        'total_revenue': total_revenue,
        'total_labor_cost': total_labor_cost,
        'total_expense_cost': total_expense_cost,
        'total_margin_dollars': total_margin_dollars,
        'overall_margin_percent': overall_margin_percent,
        'sga_pool': pools['sga_pool'],
        'data_pool': pools['data_pool'],
        'workplace_pool': pools['workplace_pool'],
        'revenue_center_count': len(revenue_centers),
        'cost_center_count': len(cost_centers),
        'non_revenue_client_count': len(non_revenue_clients),
    }

    # Phase 6: Save results
    logs.append("Saving results to database...")
    db.save_revenue_centers(batch_id, revenue_centers)
    db.save_cost_centers(batch_id, cost_centers)
    db.save_non_revenue_clients(batch_id, non_revenue_clients)
    db.save_hours_detail(batch_id, hours_detail)
    db.save_expenses_detail(batch_id, expense_detail)
    db.save_pools_detail(batch_id, pools, tagged_revenue)
    db.save_batch_summary(batch_id, summary, validation_results.to_json())

    logs.append("Analysis complete!")

    return {
        'success': True,
        'summary': summary,
        'validation': validation_results.to_json(),
        'logs': logs,
    }


class handler(BaseHTTPRequestHandler):
    """Vercel Python Function handler."""

    def do_POST(self):
        """Handle POST request to run analysis."""
        try:
            # Read request body
            content_length = int(self.headers.get('Content-Length', 0))
            body = self.rfile.read(content_length)
            data = json.loads(body) if body else {}

            batch_id = data.get('batchId')
            if not batch_id:
                self._send_error(400, "batchId is required")
                return

            # Initialize Supabase client
            db = SupabaseClient()

            # Get batch details
            batch = db.get_batch(batch_id)
            if not batch:
                self._send_error(404, f"Batch {batch_id} not found")
                return

            # Verify all files are present
            required_files = [
                'proforma_file_path',
                'compensation_file_path',
                'hours_file_path',
                'expenses_file_path',
                'pnl_file_path',
            ]
            missing = [f for f in required_files if not batch.get(f)]
            if missing:
                self._send_error(400, f"Missing file paths: {', '.join(missing)}")
                return

            # Update status to processing
            db.update_batch_status(batch_id, 'processing')

            # Run analysis
            result = run_analysis(db, batch)

            # Send success response
            self._send_json(200, result)

        except ValueError as e:
            # Business logic error
            error_msg = str(e)
            try:
                db = SupabaseClient()
                db.update_batch_status(batch_id, 'failed', error_msg)
            except Exception:
                pass
            self._send_error(400, error_msg)

        except Exception as e:
            # Unexpected error
            error_msg = f"Internal error: {str(e)}"
            error_trace = traceback.format_exc()
            print(f"MPA Process Error: {error_trace}")
            try:
                db = SupabaseClient()
                db.update_batch_status(batch_id, 'failed', error_msg)
            except Exception:
                pass
            self._send_error(500, error_msg)

    def do_OPTIONS(self):
        """Handle CORS preflight."""
        self.send_response(200)
        self._send_cors_headers()
        self.end_headers()

    def _send_json(self, status: int, data: dict):
        """Send JSON response."""
        self.send_response(status)
        self.send_header('Content-Type', 'application/json')
        self._send_cors_headers()
        self.end_headers()
        self.wfile.write(json.dumps(data).encode())

    def _send_error(self, status: int, message: str):
        """Send error response."""
        self._send_json(status, {'success': False, 'error': message})

    def _send_cors_headers(self):
        """Add CORS headers."""
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type, Authorization')

File: src/lib/mpa/process.ts
/**
 * MPA Processing Pipeline
 *
 * Main entry point for running the Monthly Performance Analysis
 */

import { createClient, SupabaseClient } from '@supabase/supabase-js';
import type {
  RevenueCenter,
  CostCenter,
  NonRevenueClient,
  HoursDetail,
  ExpenseDetail,
  OverheadPools,
  AnalysisSummary,
  ValidationItem,
} from './types';
import {
  loadProForma,
  loadCompensation,
  loadHarvestHours,
  loadHarvestExpenses,
  loadPnL,
} from './loaders';
import { classifyAllActivity } from './classification';
import {
  calculateLaborCosts,
  calculateExpenseCosts,
  mergeDirectCostsToRevenueCenters,
  calculateCostCenterCosts,
  calculateNonRevenueClientCosts,
} from './computations';
import {
  calculatePools,
  allocateSGA,
  allocateData,
  allocateWorkplace,
  calculateMargins,
  getTaggedRevenue,
} from './allocations';
import {
  runAllValidations,
  toValidationItems,
  getSummary,
  ValidationData,
} from './validators';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type AnySupabaseClient = SupabaseClient<any, any, any>;

function getSupabaseClient(): AnySupabaseClient {
  const url = process.env.NEXT_PUBLIC_SUPABASE_URL;
  const key = process.env.SUPABASE_SERVICE_ROLE_KEY;

  if (!url || !key) {
    throw new Error('Supabase not configured');
  }

  return createClient(url, key, {
    auth: { autoRefreshToken: false, persistSession: false },
  });
}

export interface ProcessingResult {
  success: boolean;
  summary: AnalysisSummary;
  validation: ValidationItem[];
  logs: string[];
}

interface BatchRecord {
  id: string;
  month_name: string;
  proforma_file_path: string;
  compensation_file_path: string;
  hours_file_path: string;
  expenses_file_path: string;
  pnl_file_path: string;
}

/**
 * Download file from Supabase Storage
 */
async function downloadFile(supabase: AnySupabaseClient, filePath: string): Promise<Buffer> {
  const { data, error } = await supabase.storage
    .from('uploads')
    .download(filePath);

  if (error) {
    throw new Error(`Failed to download ${filePath}: ${error.message}`);
  }

  const arrayBuffer = await data.arrayBuffer();
  return Buffer.from(arrayBuffer);
}

/**
 * Run the full MPA analysis pipeline
 */
export async function runAnalysis(batchId: string): Promise<ProcessingResult> {
  const logs: string[] = [];
  const supabase = getSupabaseClient();

  // Get batch record
  const { data: batch, error: batchError } = await supabase
    .from('mpa_analysis_batches')
    .select('*')
    .eq('id', batchId)
    .single();

  if (batchError || !batch) {
    throw new Error(`Batch ${batchId} not found`);
  }

  const batchRecord = batch as BatchRecord;
  const month = batchRecord.month_name;

  logs.push(`Loading files for ${month}...`);

  // Phase 1: Download and load files
  const [proformaBuffer, compBuffer, hoursBuffer, expensesBuffer, pnlBuffer] = await Promise.all([
    downloadFile(supabase, batchRecord.proforma_file_path),
    downloadFile(supabase, batchRecord.compensation_file_path),
    downloadFile(supabase, batchRecord.hours_file_path),
    downloadFile(supabase, batchRecord.expenses_file_path),
    downloadFile(supabase, batchRecord.pnl_file_path),
  ]);

  const proformaResult = loadProForma(proformaBuffer, month);
  logs.push(...proformaResult.logs);

  const compResult = loadCompensation(compBuffer);
  logs.push(...compResult.logs);

  const hoursResult = loadHarvestHours(hoursBuffer, month);
  logs.push(...hoursResult.logs);

  const expensesResult = loadHarvestExpenses(expensesBuffer);
  logs.push(...expensesResult.logs);

  const pnlResult = loadPnL(pnlBuffer);
  logs.push(...pnlResult.logs);

  logs.push('Files loaded successfully');

  // Phase 2: Classify projects
  logs.push('Classifying projects...');
  const classified = classifyAllActivity(
    proformaResult.data,
    hoursResult.data,
    expensesResult.data
  );

  logs.push(
    `Classified: ${classified.revenueCenters.length} revenue centers, ` +
    `${classified.costCenters.length} cost centers, ` +
    `${classified.nonRevenueClients.length} non-revenue clients`
  );

  // Phase 3: Compute direct costs
  logs.push('Computing direct costs...');

  const laborResult = calculateLaborCosts(hoursResult.data, compResult.data);
  logs.push(...laborResult.logs);

  const expenseResult = calculateExpenseCosts(expensesResult.data);

  let revenueCenters = mergeDirectCostsToRevenueCenters(
    classified.revenueCenters,
    laborResult.laborSummary,
    expenseResult.expenseSummary
  );

  let costCenters = calculateCostCenterCosts(
    classified.costCenters,
    laborResult.hoursDetail,
    expenseResult.expenseDetail
  );

  let nonRevenueClients = calculateNonRevenueClientCosts(
    classified.nonRevenueClients,
    laborResult.hoursDetail,
    expenseResult.expenseDetail
  );

  logs.push('Direct costs computed');

  // Phase 4: Allocate overhead
  logs.push('Allocating overhead pools...');

  const pools = calculatePools(pnlResult.data, costCenters, true);

  revenueCenters = allocateSGA(revenueCenters, pools.sgaPool);
  revenueCenters = allocateData(revenueCenters, pools.dataPool);
  revenueCenters = allocateWorkplace(revenueCenters, pools.workplacePool);
  revenueCenters = calculateMargins(revenueCenters);

  const taggedRevenue = getTaggedRevenue(revenueCenters);

  logs.push(
    `Pools allocated: SG&A $${pools.sgaPool.toLocaleString()}, ` +
    `Data $${pools.dataPool.toLocaleString()}, ` +
    `Workplace $${pools.workplacePool.toLocaleString()}`
  );

  // Phase 5: Validate
  logs.push('Running validation checks...');

  const validationData: ValidationData = {
    revenueCenters,
    costCenters,
    nonRevenueClients,
    proforma: proformaResult.data,
    pools,
    hours: hoursResult.data,
    expenses: expensesResult.data,
    compensation: compResult.data,
    pnl: pnlResult.data,
  };

  const validationResult = runAllValidations(validationData);
  logs.push(`Validation: ${getSummary(validationResult)}`);

  // Calculate summary metrics
  const totalRevenue = revenueCenters.reduce((sum, rc) => sum + rc.revenue, 0);
  const totalLaborCost = revenueCenters.reduce((sum, rc) => sum + rc.laborCost, 0);
  const totalExpenseCost = revenueCenters.reduce((sum, rc) => sum + rc.expenseCost, 0);
  const totalMarginDollars = revenueCenters.reduce((sum, rc) => sum + rc.marginDollars, 0);
  const overallMarginPercent = totalRevenue > 0 ? (totalMarginDollars / totalRevenue) * 100 : 0;

  const summary: AnalysisSummary = {
    totalRevenue,
    totalLaborCost,
    totalExpenseCost,
    totalMarginDollars,
    overallMarginPercent,
    sgaPool: pools.sgaPool,
    dataPool: pools.dataPool,
    workplacePool: pools.workplacePool,
    revenueCenterCount: revenueCenters.length,
    costCenterCount: costCenters.length,
    nonRevenueClientCount: nonRevenueClients.length,
  };

  // Phase 6: Save results
  logs.push('Saving results to database...');

  await saveResults(
    supabase,
    batchId,
    revenueCenters,
    costCenters,
    nonRevenueClients,
    laborResult.hoursDetail,
    expenseResult.expenseDetail,
    pools,
    taggedRevenue,
    summary,
    toValidationItems(validationResult)
  );

  logs.push('Analysis complete!');

  return {
    success: true,
    summary,
    validation: toValidationItems(validationResult),
    logs,
  };
}

/**
 * Save all results to database
 */
async function saveResults(
  supabase: AnySupabaseClient,
  batchId: string,
  revenueCenters: RevenueCenter[],
  costCenters: CostCenter[],
  nonRevenueClients: NonRevenueClient[],
  hoursDetail: HoursDetail[],
  expenseDetail: ExpenseDetail[],
  pools: OverheadPools,
  taggedRevenue: { totalRevenue: number; dataTaggedRevenue: number; wellnessTaggedRevenue: number },
  summary: AnalysisSummary,
  validation: ValidationItem[]
): Promise<void> {
  // Save revenue centers
  if (revenueCenters.length > 0) {
    const rcRows = revenueCenters.map(rc => ({
      batch_id: batchId,
      contract_code: rc.contractCode,
      project_name: rc.projectName,
      proforma_section: rc.proformaSection,
      analysis_category: rc.analysisCategory,
      allocation_tag: rc.allocationTag,
      revenue: rc.revenue,
      hours: rc.hours,
      labor_cost: rc.laborCost,
      expense_cost: rc.expenseCost,
      sga_allocation: rc.sgaAllocation,
      data_allocation: rc.dataAllocation,
      workplace_allocation: rc.workplaceAllocation,
      margin_dollars: rc.marginDollars,
      margin_percent: rc.marginPercent,
    }));

    const { error } = await supabase.from('mpa_revenue_centers').insert(rcRows);
    if (error) throw new Error(`Failed to save revenue centers: ${error.message}`);
  }

  // Save cost centers
  if (costCenters.length > 0) {
    const ccRows = costCenters.map(cc => ({
      batch_id: batchId,
      contract_code: cc.contractCode,
      description: cc.description,
      pool: cc.pool,
      hours: cc.hours,
      labor_cost: cc.laborCost,
      expense_cost: cc.expenseCost,
      total_cost: cc.totalCost,
    }));

    const { error } = await supabase.from('mpa_cost_centers').insert(ccRows);
    if (error) throw new Error(`Failed to save cost centers: ${error.message}`);
  }

  // Save non-revenue clients
  if (nonRevenueClients.length > 0) {
    const nrcRows = nonRevenueClients.map(nrc => ({
      batch_id: batchId,
      contract_code: nrc.contractCode,
      project_name: nrc.projectName,
      hours: nrc.hours,
      labor_cost: nrc.laborCost,
      expense_cost: nrc.expenseCost,
      total_cost: nrc.totalCost,
    }));

    const { error } = await supabase.from('mpa_non_revenue_clients').insert(nrcRows);
    if (error) throw new Error(`Failed to save non-revenue clients: ${error.message}`);
  }

  // Save hours detail
  if (hoursDetail.length > 0) {
    const hdRows = hoursDetail.map(hd => ({
      batch_id: batchId,
      contract_code: hd.contractCode,
      staff_key: hd.staffKey,
      hours: hd.hours,
      hourly_cost: hd.hourlyCost,
      labor_cost: hd.laborCost,
    }));

    const { error } = await supabase.from('mpa_hours_detail').insert(hdRows);
    if (error) throw new Error(`Failed to save hours detail: ${error.message}`);
  }

  // Save expense detail
  if (expenseDetail.length > 0) {
    const edRows = expenseDetail.map(ed => ({
      batch_id: batchId,
      contract_code: ed.contractCode,
      expense_category: null,
      notes: ed.notes,
      amount: ed.amount,
      is_billable: false,
    }));

    const { error } = await supabase.from('mpa_expenses_detail').insert(edRows);
    if (error) throw new Error(`Failed to save expense detail: ${error.message}`);
  }

  // Save pools detail (single row per batch with all pool values)
  const poolsRow = {
    batch_id: batchId,
    sga_from_pnl: pools.sgaFromPnl,
    data_from_pnl: pools.dataFromPnl,
    workplace_from_pnl: pools.workplaceFromPnl,
    nil_excluded: pools.nilExcluded || 0,
    sga_from_cc: pools.sgaFromCc,
    data_from_cc: pools.dataFromCc,
    total_revenue: taggedRevenue.totalRevenue,
    data_tagged_revenue: taggedRevenue.dataTaggedRevenue,
    wellness_tagged_revenue: taggedRevenue.wellnessTaggedRevenue,
  };

  const { error: poolsError } = await supabase.from('mpa_pools_detail').insert([poolsRow]);
  if (poolsError) throw new Error(`Failed to save pools detail: ${poolsError.message}`);

  // Update batch with summary
  const validationPassed = validation.filter(v => v.type === 'fail').length === 0;

  const { error: updateError } = await supabase
    .from('mpa_analysis_batches')
    .update({
      status: 'completed',
      total_revenue: summary.totalRevenue,
      total_labor_cost: summary.totalLaborCost,
      total_expense_cost: summary.totalExpenseCost,
      total_margin_dollars: summary.totalMarginDollars,
      overall_margin_percent: summary.overallMarginPercent,
      sga_pool: summary.sgaPool,
      data_pool: summary.dataPool,
      workplace_pool: summary.workplacePool,
      validation_passed: validationPassed,
      validation_errors: validation,
      processed_at: new Date().toISOString(),
    })
    .eq('id', batchId);

  if (updateError) throw new Error(`Failed to update batch: ${updateError.message}`);
}
